premailer/tests/test_cache.py

```diff
@@ -0,0 +1,52 @@
+from __future__ import absolute_import, unicode_literals
+
+import os
+import unittest
+import imp
+
+import cachetools
+
+
+class TestFunctionCache(unittest.TestCase):
+    def tearDown(self):
+        for key in (
+            "PREMAILER_CACHE",
+            "PREMAILER_CACHE_MAXSIZE",
+            "PREMAILER_CACHE_TTL",
+        ):
+            try:
+                del os.environ[key]
+            except KeyError:
+                pass
+
+    def test_unknown_cache(self):
+        os.environ["PREMAILER_CACHE"] = "UNKNOWN"
+
+        with self.assertRaises(Exception) as assert_context:
+            cache_fd = open("premailer/cache.py", "r")
+            imp.load_source("cache.py", "premailer/cache.py", cache_fd)
+
+        self.assertTrue(
+            assert_context.exception.args[0].startswith(
+                "Unsupported cache implementation"
+            )
+        )
+
+    def test_default_cache(self):
+        cache_fd = open("premailer/cache.py", "r")
+        cache_module = imp.load_source("cache.py", "premailer/cache.py", cache_fd)
+
+        self.assertEquals(type(cache_module.cache), cachetools.LRUCache)
+        self.assertEquals(cache_module.cache.maxsize, 128)
+
+    def test_ttl_cache(self):
+        os.environ["PREMAILER_CACHE"] = "TTL"
+        os.environ["PREMAILER_CACHE_TTL"] = "10"
+        os.environ["PREMAILER_CACHE_MAXSIZE"] = "50"
+
+        cache_fd = open("premailer/cache.py", "r")
+        cache_module = imp.load_source("cache.py", "premailer/cache.py", cache_fd)
+
+        self.assertEquals(type(cache_module.cache), cachetools.TTLCache)
+        self.assertEquals(cache_module.cache.maxsize, 50)
+        self.assertEquals(cache_module.cache.ttl, 10)
```

Comment:
The most important change of this PR is that it enabled a thread locking mechanism but there's no unit test to lock that in (no pun intended). 

Can you, first comment out the `threading.RLock()` stuff above, then create a test that starts two threads and somehow trigger the race condition. Then uncomment the threading stuff and check that it solved the problem.

Reply:
Its really hard to reproduce this race condition with sureity without modifying the cachetools code. I was actually able to reproduce it using below code, but for that I had to modify cachetools code to insert some delay while poping items from cache here: https://github.com/tkem/cachetools/blob/master/cachetools/lru.py#L35

```python
os.environ["PREMAILER_CACHE_MAXSIZE"] = "1"

from premailer import cache

RULES_MAP = {
    "h1": "{ color: blue; }",
    "h2": "{ color: green; }",
}


@cache.function_cache()
def get_styles(rule):
    return RULES_MAP[rule]


class TestFunctionCache(unittest.TestCase):
    def test_cache_multithread_synchronization(self):
        class RuleMapper(threading.Thread):
            def run(self):
	        for rule in RULES_MAP:
                    get_styles(rule)

        threads = []
        for _ in range(2):
            threads.append(RuleMapper())

        for thread in threads:
            thread.start()

        for thread in threads:
            thread.join()
```
Below are links to couple of issues that talks about this scenario if that works.
https://github.com/tkem/cachetools/issues/80
https://github.com/tkem/cachetools/issues/82

Reply:
I have no idea how you would do it but perhaps some inspiration here: 
https://github.com/peterbe/premailer/blob/4d74656fb12e8e44683fa787ae71c0735282376b/premailer/tests/test_premailer.py#L1846

I didn't actually write that code. It was a contribution. But it did (sometimes (I think)) triggered the thread-un-safety and since the contribution that test no longer fails any more.

Reply:
Thanks for the link. I was actually able to add the test by providing a overridden implementation of cache that adds some delay during delete operation on cache, thus resulting in race condition.
The test will fail if no lock is passed to `cached` decorator.